keps/sig-apps/4368-support-managed-by-label-for-batch-jobs/README.md

```diff
@@ -0,0 +1,1098 @@
+a blocker.
+
+It would also complicate debuggability of the feature.
+
+We decide to keep the label immutable, at least for [Beta](#beta), we will
+re-evaluate the decision for [GA](#ga).
+
+### Risks and Mitigations
+
+#### Ecosystem fragmentation due to forks
+
+The mechanism to disable the main Job controller opens the door for users to
+substitute it with a fork. This may create more fragmentation in the community
+as users may prefer to use their forked controllers rather than contribute
+upstream.
+
+First, this risk, to some extent, exists even today as admins with access to the
+control plane can disable job controller by passing `--controllers=-job,*` in the manifest for
+`kube-controller-manager` (see more info on the `--controllers` flag
+[here](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/)).
+
+Second, we believe that users who had the need to fork the Job controller
+already introduced dedicated Job CRDs for their needs.
+
+#### Two controllers running at the same time on old version
+
+It is possible that one configures MultiKueue (or another project) with an
+older version of k8s which does not support the label yet. In that case
+two controllers might start running and compete with Job status updates at the
+same time.
+
+Note that an analogous situation may happen when the version of Kubernetes
+already supports the label, but the feature gate is disabled in `kube-controller-manager`.
+
+To mitigate this risk we warn about it in Kueue documentation, to use this label
+only against newer versions of Kubernetes.
+
+Finally, this risk will fade away with time as the new versions of
+Kubernetes support it.
+
+#### Debuggability
+
+With this mechanism new failure modes can occur. For example, a user may make
+a typo in the label value, or the cluster administrator may not install the
+custom controller (like MultiKueue) on that cluster.
+
+In such cases the user may not observe any progress by the job for a long time
+and my need to debug the Job.
+
+In order to allow for debugging of situations like this the Job controller will
+put a log line indicating the synchronization is delegated to another controller
+(see [implementation overview](#implementation-overview)).
+
+Additionally, re-evaluate extending the `kubectl` command-line tool
+before [GA](#ga). We could extend the command to provide useful debugging
+information with the following:
+- new `MANAGED_BY` column for `kubectl get job -owide` (possibly also without `-owide`)
+- a line in the `kubectl describe job` output, just before the list of events,
+providing a user readable information if the Job is synchronized by a custom
+controller.
+
+Alternative ideas considered were
+[a dedicated condition](#condition-to-indicated-job-is-skipped)
+and [events](#event-indicating-the-job-is-skipped).
+
+#### Custom controllers not compatible with API assumptions by CronJob
```

Comment:
@atiratree responding to your question here:

> @mimowo the KEP looks good, but I do not see any considerations regarding the cronjob. Do we support setting this label to jobs in cronjob's jobTemplate?
> 
> If we do. Are there any changes required to the cronjob controller? What happens when the managing controller of the job is unresponsive or does not exist? Or when it syncs the job's status incorrectly? Do we need any consistency checks in the cronjob controller or observability on this? And it might be worth adding some tests for this as well.

First, we want to support setting `managed-by` on the `jobTemplate`. As also discussed during the wg-batch meeting the CronJob makes some assumptions about the Job Status API, that are satisfied by the Job contoller, but they are not guaranteed to be satisfied by custom controllers, adding new potential failure modes.

I added this section and a section in Design details to propose strengthening the validation to satisfy the CronJob requirements. PTAL.

Reply:
Thanks for adding this. 

Non sure how deep we should to go into the job controller and how many constraints we should include in the validation as it may not always be future proof. But having the obvious validation and validation that CronJob depends on is a good step.